soccer/gameplay/plays/offense/adaptive_formation.py

```diff
@@ -0,0 +1,295 @@
+                                 priority=10)
+
+    def execute_dribbling(self):
+        # Grab best pass
+        self.pass_target, self.pass_score = evaluation.passing_positioning.eval_best_receive_point(
+            main.ball().pos, None, main.our_robots(),
+            AdaptiveFormation.FIELD_POS_WEIGHTS,
+            AdaptiveFormation.PASSING_WEIGHTS, False)
+
+        # Grab shot chance
+        self.shot_chance = evaluation.shooting.eval_shot(main.ball().pos)
+
+        # Recalculate dribbler pos
+        self.check_dribbling_timer += 1
+        if (self.check_dribbling_timer > self.check_dribbling_timer_cutoff):
+            self.check_dribbling_timer = 0
+            self.dribbler.pos, _ = evaluation.passing_positioning.eval_best_receive_point(
+                main.ball().pos, None, main.our_robots(),
+                AdaptiveFormation.FIELD_POS_WEIGHTS,
+                AdaptiveFormation.DRIBBLING_WEIGHTS, False)
+
+        # TODO: Get list of top X pass positions and have robots in good positions to reach them
+        # Good positions can be definied by offensive / defensive costs
+        # Offensive positions move onto the ball in the direction of the goal
+        # Defensive cover the center of the field
+
+        # Setup previous values (Basic complementary filter)
+        c = .8
+        self.prev_shot_chance = c * self.shot_chance + \
+                                (1 - c) * self.prev_shot_chance
+        self.prev_pass_score = c * self.pass_score + \
+                               (1 - c) * self.prev_pass_score
+
+    def on_exit_dribbling(self):
+        self.remove_subbehavior('dribble')
+
+    def on_enter_shooting(self):
+        # TODO: Use moving kick when completed
+        self.kick = skills.pivot_kick.PivotKick()
+        self.kick.target = constants.Field.TheirGoalSegment
+        self.kick.aim_params['desperate_timeout'] = 3
+        self.add_subbehavior(self.kick, 'kick', required=False)
+
+    def on_exit_shooting(self):
+        self.remove_subbehavior('kick')
+        self.kick = None
+
+    def on_enter_clearing(self):
+        # Line kick with chip
+        # Choose most open area / Best pass, weight forward
+        # Decrease weight on sides of field due to complexity of settling
+        self.pass_target, self.pass_score = evaluation.passing_positioning.eval_best_receive_point(
+            main.ball().pos, None, main.our_robots(),
+            AdaptiveFormation.FIELD_POS_WEIGHTS,
+            AdaptiveFormation.PASSING_WEIGHTS, False)
+
+        clear = skills.pivot_kick.PivotKick()
+        clear.target = self.pass_target
+        clear.aim_params['desperate_timeout'] = 3
+        clear.use_chipper = True
+        self.add_subbehavior(clear, 'clear', required=False)
+
+    def on_exit_clearing(self):
+        self.remove_subbehavior('clear')
+
+    def on_enter_passing(self):
+        # TODO: Use the moving recieve when finished
+        self.add_subbehavior(
```

Comment:
I would like the play to have lower priority than defense at all times, basically, making the play always prefer to give robots up to defense, rather than taking them from the defense.